In [0]:
import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.model_selection import train_test_split
import sklearn as sk
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.linear_model import LassoCV,LarsCV,LassoLarsCV
from sklearn.metrics import mean_squared_error,r2_score, mean_absolute_error,accuracy_score
import matplotlib.pyplot as plt
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
mainFile =pd.read_csv('drive/My Drive/Colab Notebooks/finalmaster-ratios.csv')
mainFile.head()

# Purchases  B01001001   B01001002  ...  B19001015  B19001016  B19001017
0           22     206252  469.226965  ...  36.440765  23.446284  21.197485
1            7      61399  486.538869  ...  33.000508  33.169741  24.792689
2            3      73170  489.859232  ...  20.125056  11.890525  16.537397
3           94     251724  505.585483  ...  43.731067  38.851729  40.427349
4            0      37382  495.586111  ...  36.118530  31.603714  19.648989

[5 rows x 190 columns]

In [0]:
allvariablenames = list(mainFile.columns.values)

In [0]:
target = mainFile["# Purchases"] 
listofallpredictors = mainFile.iloc[:,8:190] 
predictors = listofallpredictors

In [0]:
pred_train, pred_test, tar_train, tar_test = train_test_split(predictors, target, test_size=.3, random_state=123)

In [0]:
model = LassoLarsCV(cv=10, precompute=False).fit(pred_train, tar_train)

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 5 iterations, i.e. alpha=1.496e+00, with an active set of 5 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 8 iterations, i.e. alpha=1.098e+00, with an active set of 8 regressors, and the smallest cholesky pivot element being 6.322e-08. Reduce max_iter or increase eps parameters.
  ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/least_angle.py:578: ConvergenceWarning: Regressors in active set degenerate. Dropping a regressor, after 10 iterations, i.e. alpha=7.329e-01, with an active set of 10 regressors, and the smallest cholesky pivot element b

In [0]:
predictors_model=pd.DataFrame(listofallpredictors.columns)
predictors_model.columns = ['label']
predictors_model['coeff'] = model.coef_

for index, row in predictors_model.iterrows():
    if row['coeff'] > 0:
        print(row.values)

['B01001014' 0.8558761066941788]
['B01001036' 2.5053482381631653]
['B01001037' 0.8892493223320962]
['B01001038' 1.5316387928880384]
['B02001005' 0.41252295298457853]
['B13014026' 0.48004105312075906]
['B13014027' 0.6978957445987839]
['B13016001' 875149895.329212]
['B19001017' 1.4834348068681533]


The first three lines of the code is extracting the column name and the coefficient value for the same column. The for loop in this code is checking if the coefficient is greater that 0 and if it is, it is printing the value.

B01001014 | Male Age 40-44 years

---


B01001036 | Female Age 30-34 years


---


B01001037 | Female Age 35-39 years


---


B01001038 | Female Age 40-44 years

---


B02001005 | Asian race

---


B13014026 | Bachelor's Degree

---


B13014027 | Graduate and Professional Degree

---


B13016001 | total Women 15 to 50 years

---


B19001017 | Household income of $200,000 or More





According to the Lasso Model, areas with Female from age 30-44, we sell more Bobo bars. Along with that, we sell more bobo bars in the area where there are male of age 40-44. We also sell more bobo bars, in the asian race. Bobo bars are also sold in the area with people with Bachelors, Graduate or Professional degrees. Along with that, bobo bars are sold mroe in the area of household with the income of $200,000 or higher. Last but not the least, the biggest market for selling bobo bars is the area with women aged from 15-50 years. 

If I had to report two census varible to my boss, I would report with women aged from 15-50 years and female with age 30-34 years old.
Since, the 1st variable already encompasses the second one, the other census varible would be Household income of $200,000 or More.


In [0]:
train_error = mean_squared_error(tar_train, model.predict(pred_train))
print ('training data MSE')
print(train_error)

training data MSE
22025.491066757


In [0]:
test_error = mean_squared_error(tar_test, model.predict(pred_test))
print ('Test data MSE')
print(test_error)

Test data MSE
41549.54803776253


The mean squared error of training set is 22025.491066757 and the mean squared error of Test set is 41549.54803776253. We have very high mean squared error for both training and test set. This means that the data is overfitted as MSE in test set is much higher that MSE in training set. 



In [0]:
rsquared_train=model.score(pred_train,tar_train)
print ('training data R-square')
print(rsquared_train)

training data R-square
0.2400221219784492


In [0]:
rsquared_test=model.score(pred_test,tar_test)
print ('Test data R-square')
print(rsquared_test)

Test data R-square
0.1758628512005107


The r-squared in both training and test set is lower which means the model doesn't fit the data. This means that the census data doesn't quite predict the sales. 

In [0]:
print("y interecept:")
print(model.intercept_)

y interecept:
22.19738813257551


Our baseline sales number is 22.19. The expected sales number when all the predictors are 0 is 22.19.